crates/aptos-crypto/src/bls12381.rs

```diff
@@ -0,0 +1,381 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides API for the BLS12-381 signature library (https://github.com/supranational/blst).
+
+use crate::{
+    hash::CryptoHash, signing_message, CryptoMaterialError, Length, PrivateKey, PublicKey,
+    Signature, SigningKey, Uniform, ValidCryptoMaterial, ValidCryptoMaterialStringExt,
+    VerifyingKey,
+};
+use anyhow::{anyhow, Result};
+use aptos_crypto_derive::{DeserializeKey, SerializeKey, SilentDebug, SilentDisplay};
+use blst::{blst_scalar, BLST_ERROR};
+use rand_core::{OsRng, RngCore};
+use serde::Serialize;
+use std::convert::TryFrom;
+
+const DST: &[u8] = b"BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_";
+
+#[derive(Clone, Eq, PartialEq, SerializeKey, DeserializeKey)]
+/// A BLS12381 public key
+pub struct BLS12381PublicKey {
+    pubkey: blst::min_sig::PublicKey,
+}
+
+#[derive(SerializeKey, DeserializeKey, SilentDebug, SilentDisplay)]
+/// A BLS12381 private key
+pub struct BLS12381PrivateKey {
+    privkey: blst::min_sig::SecretKey,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, SerializeKey, DeserializeKey)]
+/// A BLS12381 signature
+pub struct BLS12381Signature {
+    sig: blst::min_sig::Signature,
+}
+
+impl BLS12381PublicKey {
+    /// The length of the BLS12381PublicKey
+    pub const LENGTH: usize = 96;
+
+    /// Serialize a BLS12381PublicKey.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.pubkey.to_bytes()
+    }
+
+    /// Validate the public key.
+    pub fn validate(&self) -> Result<()> {
+        self.pubkey.validate().map_err(|e| anyhow!("{:?}", e))
+    }
+}
+
+impl BLS12381PrivateKey {
+    /// The length of the BLS12381PrivateKey
+    pub const LENGTH: usize = 32;
+
+    /// Serialize a BLS12381PrivateKey.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.privkey.to_bytes()
+    }
+}
+
+impl BLS12381Signature {
+    /// The length of the BLS12381Signature
+    pub const LENGTH: usize = 48;
+
+    /// Serialize a BLS12381Signature.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.sig.to_bytes()
+    }
+
+    /// Validate the signature.
+    pub fn validate(&self) -> Result<()> {
+        self.sig.validate(true).map_err(|e| anyhow!("{:?}", e))
+    }
+
+    /// Aggregate multiple signatures, assume individual signature are validated.
+    pub fn aggregate(sigs: Vec<Self>) -> Result<Self> {
+        let sigs: Vec<_> = sigs.iter().map(|s| &s.sig).collect();
+        let agg_sig = blst::min_sig::AggregateSignature::aggregate(&sigs[..], false)
```

Comment:
I am wondering if it's safe to aggregate without group checking the signature shares. Specifically, maybe there's an attack where one could craft a signature share outside of the group that makes the aggregated multisignature verify.

Reply:
I think if we verify the multisignature after aggregation with the `Signature::verify` API, which we have to do anyway, then we should be good: the `Signature::verify` API will group-check the final multisignature. If an attacker is able to craft an adversarial signature share that when input into this `aggregate` function leads to a forgery in the multisignature scheme, then that attacker has broken the scheme anyway, since the attacker model does not assume an "honest" aggregate function.